### PR TITLE
Provider service factory

### PR DESCRIPTION
**Pull request by SimonRichardson**
Implements the provider service factory worker along with cloud
and credential provider services. The provider services aren't
just regular services, instead, they're a subset of methods that we
offer. If the return type of the service factory was an interface
we could just have a different interface, but we use concrete types,
so we need to re-encapsulate that logic. Offering out a full service
could give developers the wrong idea in the future, that the
service factory and provider service factory could be used
interchangeably. This would be a miss-step, we want to ensure they're
different types, so that nobody accidentally depends on them.

The services themselves offer out what the full-service offers, but
that doesn't always have to be the case, we could diverge as they do
satisfy other requirements.

Most of this is a clone from the service-factory worker code, so it's
reuse is heavily inspired here.

----

_EDIT_: I've also added ModelConfig to the Provider service.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ juju bootstrap lxd test
```

## Links

**Jira card:** JUJU-5642

### CODE REVIEW

**Review by nvinuesa — COMMENTED** [review]
Some questions about naming and signatures

File: internal/worker/providerservicefactory/manifold.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	Warningf(message string, args ...any)
+}
+
+// ManifoldConfig holds the information necessary to run a provider service
+// factory worker in a dependency.Engine.
+type ManifoldConfig struct {
+	ChangeStreamName                string
+	Logger                          Logger
+	NewWorker                       func(Config) (worker.Worker, error)
+	NewProviderServiceFactoryGetter ProviderServiceFactoryGetterFn
```
Comment (manadart) [inline internal/worker/providerservicefactory/manifold.go:34]: Good to comment these here in the in the worker, particularly regarding one being for supply to consumers.

**Review by manadart — COMMENTED** [review]
Looking pretty good.

File: cmd/jujud-controller/agent/machine/manifolds.go [hunk 2]
```diff
@@ -717,14 +718,22 @@ func commonManifolds(config ManifoldsConfig) dependency.Manifolds {
...
-			NewControllerServiceFactory: workersf.NewControllerServiceFactory,
-			NewModelServiceFactory:      workersf.NewModelServiceFactory,
+			Logger:                      workerservicefactory.NewLogger("juju.worker.servicefactory"),
+			NewWorker:                   workerservicefactory.NewWorker,
+			NewServiceFactoryGetter:     workerservicefactory.NewServiceFactoryGetter,
+			NewControllerServiceFactory: workerservicefactory.NewControllerServiceFactory,
+			NewModelServiceFactory:      workerservicefactory.NewModelServiceFactory,
+		}),
+
+		providerServiceFactoryName: providerservicefactory.Manifold(providerservicefactory.ManifoldConfig{
```
[the remark this answers is not in the document]
Comment (SimonRichardson) [reply cmd/jujud-controller/agent/machine/manifolds.go:731]: Correct. That'll come later.

**Review by manadart — APPROVED** [review]
Thanks. Looks good now.

**Pull request merged**